engine/lite/kafka/runtime/src/main/scala/pl/touk/nussknacker/engine/lite/kafka/KafkaProducerRecordsHandler.scala

```diff
@@ -0,0 +1,125 @@
+      new ManuallyCommittingProducerRecordsHandler(producer)
+    }
+  }
+
+}
+
+trait KafkaProducerRecordsHandler extends AutoCloseable {
+
+  protected def producer: KafkaProducer[Array[Byte], Array[Byte]]
+
+  def enrichConsumerProperties(consumerProperties: Properties): Unit
+
+  def beforeRecordsProcessing(): Unit
+
+  def onRecordsSuccessfullyProcessed(records: ConsumerRecords[Array[Byte], Array[Byte]], consumer: KafkaConsumer[_, _]): Unit
+
+  def onRecordsProcessingFailure(): Unit
+
+  final def send(record: ProducerRecord[Array[Byte], Array[Byte]]): Future[RecordMetadata] = {
+    KafkaUtils.sendToKafka(record)(producer)
+  }
+
+  final def metrics(): util.Map[MetricName, _ <: Metric] =
+    producer.metrics()
+
+  override def close(): Unit = {
+    producer.close()
+  }
+
+}
+
+class TransactionalProducerRecordsHandler(protected val producer: KafkaProducer[Array[Byte], Array[Byte]], consumerGroupId: String)
+  extends KafkaProducerRecordsHandler {
+
+  override def enrichConsumerProperties(consumerProperties: Properties): Unit = {
+    // default is read uncommitted which is not a good default for transactions
+    KafkaUtils.setIsolationLevelIfAbsent(consumerProperties, IsolationLevel.READ_COMMITTED)
+    // offset commit is done manually in sendOffsetsToTransaction
+    consumerProperties.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)
+  }
+
+  override def beforeRecordsProcessing(): Unit = {
+    producer.beginTransaction()
+  }
+
+  @silent("deprecated")
+  override def onRecordsSuccessfullyProcessed(records: ConsumerRecords[Array[Byte], Array[Byte]], consumer: KafkaConsumer[_, _]): Unit = {
+    val offsetsMap: Map[TopicPartition, OffsetAndMetadata] = retrieveMaxOffsetsOffsets(records)
+    // group metadata commit API requires brokers to be on version 2.5 or above so for now we use deprecated api
+    producer.sendOffsetsToTransaction(offsetsMap.asJava, consumerGroupId)
+    producer.commitTransaction()
+  }
+
+  override def onRecordsProcessingFailure(): Unit = {
+    producer.abortTransaction()
+  }
+
+  private def retrieveMaxOffsetsOffsets(records: ConsumerRecords[Array[Byte], Array[Byte]]): Map[TopicPartition, OffsetAndMetadata] = {
+    records.iterator().asScala.map { rec =>
+      val upcomingOffset = rec.offset() + 1
+      (new TopicPartition(rec.topic(), rec.partition()), upcomingOffset)
+    }.toList.groupBy(_._1).mapValuesNow(_.map(_._2).max).mapValuesNow(new OffsetAndMetadata(_))
+  }
+
+}
+
+object TransactionalProducerRecordsHandler {
+
+  def apply(producerProperties: Properties, consumerGroupId: String): TransactionalProducerRecordsHandler = {
+    //FIXME generate correct id - how to connect to topic/partition??
+    producerProperties.put("transactional.id", consumerGroupId + UUID.randomUUID().toString)
+    val producer = new KafkaProducer[Array[Byte], Array[Byte]](producerProperties)
+    producer.initTransactions()
+    new TransactionalProducerRecordsHandler(producer, consumerGroupId)
```

Comment:
maybe constructor should be private? So it's only possible to create this instance using this apply method, which calls `initTransactions` which is obligatory?

Reply:
Good point

Reply:
Done